text/0000-dynamic-attributes.md

```diff
@@ -0,0 +1,91 @@
+---
+title: Passing Dynamic Attributes/Properties
+status: DRAFTED
+created_at: 2021-07-05
+updated_at: YYYY-MM-DD
+champion: Mohammed Abdul Sattar (abdulsattar)
+pr: https://github.com/salesforce/lwc-rfcs/pull/52
+---
+
+# Passing Dynamic Attributes/Properties
+
+## Summary
+This proposal adds mechanism to pass a dynamic attributes or properties bag to elements in
+an LWC template.
+
+## Example
+```js
+// x/myComponent.js
+export default class MyComponent extends LightningElement {
+    childProps = { 
+        name: 'My Child',
+        style: 'width: 50px',
+        class: 'my-class',
+        onclick: this.handleClick,
+    };
+    
+    handleClick() {
+        console.log('Clicked')
+    }
+}
+```
+```html
+// x/myComponent.html
+<template>
+    <x-child lwc:spread={childProps}></x-child>
+</template>
+```
+
+The new `lwc:spread` directive passes `childProps` as properties/attributes to `x-child`.
+
+## Motivation
+
+In LWC, there is no way to pass props/attributes to a child component when the component author
+doesn't know in advance all the props/attributes that are needed.
+Adding this feature can be helpful in creating wrapper components that extend the functionality
+of an existing component while being oblivious to all the props that component requires.
+
+### Prior Art
+- React ([JSX Spread Operator](https://reactjs.org/docs/jsx-in-depth.html#spread-attributes))
+- Vue ([`v-bind`](https://v3.vuejs.org/api/directives.html#v-bind))
+- Svelte ([Spread Props](https://svelte.dev/tutorial/spread-props))
+- Lit ([Discussion for support](https://github.com/lit/lit/issues/923))
+
+## Detailed Design
+
+A new LWC directive, `lwc:spread` will be introduced that can be used to spread the props/attributes
+to the children elements.
+
+### Structure
+It accepts an object with keys as the attribute names and values as the
+corresponding attribute values.
+
+LWC Template compiler separates all attributes into:
+1. Attributes
+2. Properties
+3. `style` attribute
+4. `class` attribute
+
+The same algorithm will be used for separating those attributes at runtime.
```

Comment:
Just to confirm, the passed object is a flat object and the LWC engine will have to figure out at runtime if an object property is either an attribute, a property, or an event listener. Correct?

A couple of questions on this topic:
- What happens when a property is present in a rendering cycle and removed in the next rendering cycle (eg. `{ name: 'foo' }` --> `{ }`)? What does it mean for properties, attributes, and event listeners?
- What do `null` and `undefined` map to?
- Until now all the event listeners are automatically bound to the component (`this` value being the component instance)? What is the expected behavior with this proposal?

Reply:
> Just to confirm, the passed object is a flat object and the LWC engine will have to figure out at runtime if an object property is either an attribute, a property, or an event listener. Correct?

Correct.

Answers to your questions:
- Attributes and event listeners are removed. Not clear what to do with properties. I put it as an open question.
- Maybe we can use `null` and `undefined` to remove the properties?
- We should keep the same behavior.

Reply:
I was curious what the other frameworks do here. In the case of Svelte, Vue, and React, they all seem to agree that

```js
{
  foo: '',
  bar: null,
  baz: undefined
}
```

should render as `<div foo=""></div>`. This makes sense to me.

![Screen Shot 2021-07-06 at 12 12 57 PM](https://user-images.githubusercontent.com/283842/124654879-f80a2b80-de53-11eb-80eb-0fcdee64f89c.png)
![Screen Shot 2021-07-06 at 12 02 23 PM](https://user-images.githubusercontent.com/283842/124654885-f9d3ef00-de53-11eb-9cbc-cebd85e18d09.png)
![Screen Shot 2021-07-06 at 12 10 17 PM](https://user-images.githubusercontent.com/283842/124654889-fa6c8580-de53-11eb-8d05-d98ce46bbbda.png)

Reply:
I looked at what LWC does currently with `null`, `undefined`. I'd prefer if we keep the same behavior.

![image](https://user-images.githubusercontent.com/84044/124709855-9d3b0780-df19-11eb-8ce9-e86841375d35.png)

Reply:
As a side note, Svelte uses runtime reflection to know if a value should be set as property or as attribute: https://github.com/sveltejs/svelte/blob/b46432060753c4fa3db11da8641f67e49c0016ce/src/runtime/internal/dom.ts#L239

Reply:
> For removed properties, I don't think we need to do anything special. Unlike all the frameworks listed above, in LWC, properties are passed individually and there is no guarantee a given property is passed or not via the template. In this regard, I don't think components should be notified that a property present in the previous rendering cycle is not present anymore.

The issue with this is:
```
<lightning-input lwc:spread={inputProps}></x-img>
```
If we start with `inputProps` with `{}`, then add `{disabled: true}` and then set it back to `{}`, we are saying `disabled` will be `true`.

We also can't say we'll explicitly mark all the missing props in the next cycle `undefined` because, when we do set `inputProps` to `{}`, the initial value of `disabled` might be `false` (as declared in the JS class). If we later set it to `{disabled: true}` and then to `{}` (and LWC marks it as `undefined`), it's inconsistent.

Reply:
You are right, we should make it super explicit in the documentation that LWC won't do any magic with props. Unlike attributes, there is no way to know what is the initial value for a given prop.